### PR TITLE
[workflow] Add a 'veristat' job

### PR DESCRIPTION
**Pull request by eddyz87**
Adds a job to execute veristat on a selected set of object files as a part of the general test pipeline. The job is executed differently for pushes and for pull requests:

- For pushes:
  - the kernel is compiled as usual;
  - selftests are compiled as well, this is necessary to produce object files that would be used by veristat;
  - veristat is executed inside qemu, performance statistics is collected and cached as baseline data.

- For pull requests:
  - veristat is executed inside qemu and performance statistics is collected;
  - baseline data is restored from cache;
  - baseline data is compared to current statistics and step summary page is formed;
  - if there are some programs that verified successfully in the baseline but fail verification in current run, the step is marked as failed.

This also adds 'run_veristat' "entrypoint" for run_selftests.sh.

This entrypoint executes veristat on a set of binary files specified in ./tools/testing/selftests/bpf/veristat.cfg file and saves ouput to the /command_output/ directory within qemu image. This output is extracted from the image by run-qemu action.

This is a followup for [this](https://github.com/kernel-patches/vmtest/pull/215) pull request, which was reverted. This one is changed to not fail `veristat` job if baseline is not available. Here is a link to a test execution under such conditions:

https://github.com/eddyz87/bpf/actions/runs/4758707630/jobs/8457542575

### CODE REVIEW

**Review by danielocfb — APPROVED** [review]
Looks good to me, thanks Eduard!

File: .github/scripts/veristat-compare.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+STATES_NEW  = 6
+
+if len(sys.argv) == 2:
+    CSV_PATH = sys.argv[1]
+else:
+    print(f'Usage: {sys.argv[0]} <compare.csv>>')
+    sys.exit(1)
+
+if not os.path.exists(CSV_PATH):
+    print(f'# {CSV_PATH} does not exist, failing step')
```
Comment (chantra) [inline .github/scripts/veristat-compare.py:41]: any reasons to print to stdout vs stderr? or not using the logging module?

File: .github/scripts/veristat-compare.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        if headers != HEADERS:
+            raise Exception(f'Unexpected table header for {path}: {headers}')
+        return list(reader)
+
+table = read_table(CSV_PATH)
+new_failures = False
+changes = False
+headers = ['File', 'Program', 'Verdict', 'States Diff (%)']
+html_table = [headers]
+text_table = [headers]
```
Comment (chantra) [inline .github/scripts/veristat-compare.py:57]: looks like those are global variables. Could we limit the use of globals and favor passing arguments or objects to abstract it?

File: .github/scripts/veristat-compare.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+HEADERS = ['file_name', 'prog_name', 'verdict_base', 'verdict_comp',
+           'verdict_diff', 'total_states_base', 'total_states_comp',
+           'total_states_diff']
+
+FILE        = 0
+PROG        = 1
+VERDICT_OLD = 2
+VERDICT_NEW = 3
+STATES_OLD  = 5
+STATES_NEW  = 6
```
Comment (chantra) [inline .github/scripts/veristat-compare.py:32]: This could be wrapped in an [enum](https://docs.python.org/3/library/enum.html)

File: .github/scripts/veristat-compare.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+HEADERS = ['file_name', 'prog_name', 'verdict_base', 'verdict_comp',
+           'verdict_diff', 'total_states_base', 'total_states_comp',
+           'total_states_diff']
+
+FILE        = 0
+PROG        = 1
+VERDICT_OLD = 2
+VERDICT_NEW = 3
+STATES_OLD  = 5
```
Comment (chantra) [inline .github/scripts/veristat-compare.py:31]: is 4 missing on purpose?

**Comment by danielocfb** [comment]
Manu mentioned that he also wants to take a look, so will wait for him.

File: .github/scripts/veristat-compare.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+if new_failures:
+    section_name = '# There are new veristat failures'
+elif changes:
+    section_name = '# There are changes in verification performance'
+else:
+    section_name = '# No changes in verification performance'
+
+# Print the step summary
+
+if summary_path := os.getenv('GITHUB_STEP_SUMMARY', None):
```
Comment (chantra) [inline .github/scripts/veristat-compare.py:125]: TIL `:=` :D
This was introduced in python3.8 apparently. The containers have python3.8.10 so we should be fine with this syntax.

I am not sure we should silently ignore this. This script's main purpose is to provide this summary, right?

When running outside of GH actions, this can be worked around by defining the env var on the command line.

For #214 I basically attempt to document this at the top of the file with an example run one-liner: https://github.com/kernel-patches/vmtest/pull/214/files#diff-cf4f5f0c0e0169cd40d613cf183cf3d4a8b8f058e2560b3853ef6aa3717dcdfaR3

This file could probably benefit from this too?

**Review by chantra — COMMENTED** [review]
Thanks @eddyz87 !

One comment I have regarding the python script is that currently it is using `globals`, which I would rather avoid and explicitely pass parameters around.

Another thing is that it seems to mix program control flow with definitions of functions.
Also, I would rather have the main control flow gated under `if __name__ == '__main__'`.

I would favor use of argparse to manually dealing with arguments. While probably not a big deal now, it will make future changes easier, and we don't have to deal with arg parsing when it is one import away in core python.
Same for logging, let's use the logging module.
Those 2 are the canonical way to handle args, log in python, let's re-use them unless there is a good reason not to.

I think this change could benefit from shipping fixture for cases where we expect the script to report failure and when we expect it to be fine. There again, in the vain of #214, we would be able to validate:
1) we exist with the return code we expect (regression detected or not)
2) We have a simple way to confirm that the script does not explode at run time.

While not necessarily asking to hook the testing in this PR, having those fixtures, they could be used as part of the test plan.

**Comment by chantra** [comment]
TL;DR-ing my comments. This is non-blocking, but I think addressing them will make the project more sustainable. We have already accumulated a bunch of tech-debts in the past and this is making the project hard to maintain, so I would like to avoid adding more to it.

Are you onboard for addressing those comments in future diffs?

File: .github/scripts/veristat-compare.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
+import sys
```
[the remark this answers is not in the document]
Comment (eddyz87) [reply .github/scripts/veristat-compare.py:1]: Will do

File: .github/scripts/veristat-compare.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+VERDICT_OLD = 2
+VERDICT_NEW = 3
+STATES_OLD  = 5
+STATES_NEW  = 6
+
+if len(sys.argv) == 2:
+    CSV_PATH = sys.argv[1]
+else:
+    print(f'Usage: {sys.argv[0]} <compare.csv>>')
+    sys.exit(1)
```
[the remark this answers is not in the document]
Comment (eddyz87) [reply .github/scripts/veristat-compare.py:38]: Will do

File: .github/scripts/veristat-compare.py (hunk 1 above)
Comment (eddyz87) [reply .github/scripts/veristat-compare.py:41]: No reasons, just wanted to keep things simple, will change to `logging`

File: .github/scripts/veristat-compare.py (hunk 2 above)
Comment (eddyz87) [reply .github/scripts/veristat-compare.py:57]: will wrap as a function

File: .github/scripts/veristat-compare.py (hunk 3 above)
Comment (eddyz87) [reply .github/scripts/veristat-compare.py:32]: These are used as indices in a vector, using `Enum` would make the notation less readable: `row[FILE]` &rarr; `row[FILE.Value]`, so I'd rather not use enum in this case.

File: .github/scripts/veristat-compare.py (hunk 4 above)
Comment (eddyz87) [reply .github/scripts/veristat-compare.py:31]: yes, sample of the record in the comment at the [start](https://github.com/kernel-patches/vmtest/pull/218/commits/88630c9ea660724f5017584bb260ea63645a7766#diff-d9bba8d01ee121eade3ef637cc83865fac60cb310a246ae82de3fd98c6dfd33eR5) of the file, here are the columns:

```
# Columns:
#  0. file_name
#  1. prog_name
#  2. verdict_base
#  3. verdict_comp
#  4. verdict_diff
#  5. total_states_base
#  6. total_states_comp
#  7. total_states_diff
```

The `verdict_diff` is ignored.

File: .github/scripts/veristat-compare.py (hunk 5 above)
Comment (eddyz87) [reply .github/scripts/veristat-compare.py:125]: I can skip the conditional, will rely on python exiting with error if file can't be opened.

**Comment by eddyz87** [comment]
> TL;DR-ing my comments. This is non-blocking, but I think addressing them will make the project more sustainable. We have already accumulated a bunch of tech-debts in the past and this is making the project hard to maintain, so I would like to avoid adding more to it.
> 
> Are you onboard for addressing those comments in future diffs?

I can address these today as a part of this pull request.

**Comment by chantra** [comment]
Thanks @eddyz87 !

**Comment by eddyz87** [comment]
> Thanks @eddyz87 !

Just pushed the updated script, I think I covered all the above points.

**Comment by chantra** [comment]
Thanks @eddyz87 . Sorry for the late review. Let's merge this tomorrow.
